### DISCUSSION TITLE
How to scale NanoMQ to handle 500K concurrent connections?

### DISCUSSION THREAD

**Discussion by AmirDotDev0** [nanomq/nanomq#741] [Q&A] [answered]
I'm expecting 500K users, and I want to run NanoMQ on a dedicated server with the following specs:

CPU: 2x AMD EPYC 7282 - 16Cores / 32 threads.
Memory: 256 GB REG ECC - can get up to 512 GB if needed.
Network: 10 Gbit/s Port.

1. I need to know if NanoMQ is capable of handling the number of users expected?
2. I need to know if my server with the above specs can handle the number of users expected?
3. How does NanoMQ scale?

**Comment by JaylinYu** [comment, marked as answer]
Hi Osman,
NanoMQ scale well on multi-core SMP platform. I would like to call it best stand-alone MQTT broker.
I manage to sustain 120k clients with 60k msg in and 60k msg out throughput:
![image](https://user-images.githubusercontent.com/64823539/179781902-55cd3671-b8ed-4fb0-8684-74f1e6d733d6.png)
![image](https://user-images.githubusercontent.com/64823539/179781941-e293fca6-1902-4ca3-a5f7-a4b9e8c0a447.png)
![image](https://user-images.githubusercontent.com/64823539/179781980-0208443b-972e-4fca-8515-cc3229020d51.png)
![image](https://user-images.githubusercontent.com/64823539/179782009-0bdc418c-37f2-43f0-825e-aca63d420ee4.png)
Benchmark is on a 8c16g cloud server (Intel skylake 3GHZ ), with only 50 percent CPU usage.
I believe you can achieve better performance with Zen2 EPYC.

BTW I reached 2million above msg/s on Zen3 5950x.

**Reply by AmirDotDev0** [reply]
This is awesome, thanks.

**Comment by JaylinYu** [comment]
About tuning guide, please check README, most rely on taksq thread and msq_len

**Reply by AmirDotDev0** [reply]
Will do.
